osquery/tables/system/darwin/disk_encryption.mm

```diff
@@ -0,0 +1,299 @@
+  if (CFDictionaryGetValueIfPresent(
+          properties, CFSTR("efilogin-unlock-ident"), &unlock_ident)) {
+    if (CFGetTypeID(unlock_ident) != CFDataGetTypeID()) {
+      return Status(1, "Unexpected data type for unlock ident");
+    }
+    uuid = CFDataCreateCopy(kCFAllocatorDefault, (CFDataRef)unlock_ident);
+    if (uuid == nullptr) {
+      return Status(1, "Could not get UUID");
+    }
+    CFRelease(properties);
+    return Status(0, "ok");
+  }
+
+  return Status(1, "Could not get unlock ident");
+}
+
+Status genUid(id_t& uid, uuid_string_t& uuid_str) {
+  CFDataRef uuid = nullptr;
+  if (!genUnlockIdent(uuid).ok()) {
+    if (uuid != nullptr) {
+      CFRelease(uuid);
+    }
+    return Status(1, "Could not get unlock ident");
+  }
+
+  CFDataGetBytes(uuid, CFRangeMake(0, CFDataGetLength(uuid)), (UInt8*)uuid_str);
+  if (uuid != nullptr) {
+    CFRelease(uuid);
+  }
+
+  uuid_t uuidT = {0};
+  if (uuid_parse(uuid_str, uuidT) != 0) {
+    return Status(1, "Could not parse UUID");
+  }
+
+  // id_type >=0 are all valid id types
+  int id_type = -1;
+  if (mbr_uuid_to_id(uuidT, &uid, &id_type) != 0 && id_type != ID_TYPE_UID) {
+    return Status(1, "Could not get uid from uuid");
+  }
+
+  return Status(0, "ok");
+}
+
+void genFDEStatusForBSDName(const std::string& bsd_name,
+                            const std::string& uuid,
+                            bool isAPFS,
+                            QueryData& results) {
+
+  auto matching_dict =
+      IOBSDNameMatching(kIOMasterPortDefault, kNilOptions, bsd_name.c_str());
+  if (matching_dict == nullptr) {
+    return;
+  }
+
+  auto service =
+      IOServiceGetMatchingService(kIOMasterPortDefault, matching_dict);
+  if (!service) {
+    return;
+  }
+
+  CFMutableDictionaryRef properties;
+  if (IORegistryEntryCreateCFProperties(
+          service, &properties, kCFAllocatorDefault, kNilOptions) !=
+      KERN_SUCCESS) {
+    IOObjectRelease(service);
+    return;
+  }
+
+  Row r;
+  r["name"] = kDeviceNamePrefix + bsd_name;
+  r["uuid"] = uuid;
+
+  if (isAPFS) {
+    auto bundle_url = CFURLCreateWithFileSystemPath(
+      kCFAllocatorDefault,
+      CFSTR("/System/Library/PrivateFrameworks/DiskManagement.framework"),
+      kCFURLPOSIXPathStyle,
+      true);
+    if (bundle_url == nullptr) {
+      LOG(ERROR) << "parsing DiskManagement bundle URL";
+    }
+
+    auto bundle = CFBundleCreate(kCFAllocatorDefault, bundle_url);
+    CFRelease(bundle_url);
+    if (bundle == nullptr) {
+      LOG(ERROR) << "opening DiskManagement bundle";
+   }
+
+    CFBundleLoadExecutable(bundle);
+
+    DASessionRef session = DASessionCreate(kCFAllocatorDefault);
+
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Warc-performSelector-leaks"
+    // DMManager *man = [DMManager sharedManager]
+    id cls = NSClassFromString(@"DMManager");
+    if (![cls respondsToSelector:@selector(sharedManager)]) {
+      LOG(ERROR) << "DMManager does not respond to sharedManager selector";
```

Comment:
I could use some advice as to how to structure the error handling here and below. We need to be careful to make the appropriate `CFRelease` calls depending on how many of the calls that perform initialization succeeded. We can't rely on ARC or RAII to do this cleanup (from what I can tell) because of the dynamic nature of the calls that are made.

Maybe something like the ScopeGuard with `std::function` pattern?

Reply:
[Background on ScopeGuard](https://stackoverflow.com/a/31365171/491710)

Reply:
@marpaia thinks I should do the appropriate releases manually in each location and leave a comment imploring future developers to beware. I'm inclined to agree so that we can move on.

Reply:
The PR is now updated to do error handling and manual cleanup.